src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -397,27 +370,426 @@ CollectionExpressionTypeKind.Span or
+            {
+                Debug.Assert(useKnownLength);
+                Debug.Assert(setCount is { });
+                Debug.Assert(asSpan is { });
+
+                // CollectionsMarshal.SetCount<ElementType>(list, N + s1.Length + ...);
+                sideEffects.Add(_factory.Call(receiver: null, setCount, listTemp, GetKnownLengthExpression(elements, numberIncludingLastSpread, localsBuilder)));
+
+                // var span = CollectionsMarshal.AsSpan<ElementType(list);
+                BoundLocal spanTemp = _factory.StoreToTemp(_factory.Call(receiver: null, asSpan, listTemp), out assignmentToTemp, isKnownToReferToTempIfReferenceType: true);
+                localsBuilder.Add(spanTemp);
+                sideEffects.Add(assignmentToTemp);
+
+                // Populate the span.
+                var spanGetItem = ((MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_Span_T__get_Item)).AsMember((NamedTypeSymbol)spanTemp.Type);
+
+                // int index = 0;
+                BoundLocal indexTemp = _factory.StoreToTemp(
+                    _factory.Literal(0),
+                    out assignmentToTemp,
+                    isKnownToReferToTempIfReferenceType: true);
+                localsBuilder.Add(indexTemp);
+                sideEffects.Add(assignmentToTemp);
+
+                AddCollectionExpressionElements(
+                    elements,
+                    spanTemp,
+                    localsBuilder,
+                    numberIncludingLastSpread,
+                    sideEffects,
+                    (ArrayBuilder<BoundExpression> expressions, BoundExpression spanTemp, BoundExpression rewrittenValue) =>
+                    {
+                        Debug.Assert(spanTemp.Type is NamedTypeSymbol);
+                        Debug.Assert(indexTemp.Type is { SpecialType: SpecialType.System_Int32 });
+
+                        var expressionSyntax = rewrittenValue.Syntax;
+                        var elementType = ((NamedTypeSymbol)spanTemp.Type).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0].Type;
+
+                        // span[index] = element;
+                        expressions.Add(
+                            new BoundAssignmentOperator(
+                                expressionSyntax,
+                                _factory.Call(spanTemp, spanGetItem, indexTemp),
+                                rewrittenValue,
+                                isRef: false,
+                                elementType));
+                        // index = index + 1;
+                        expressions.Add(
+                            new BoundAssignmentOperator(
+                                expressionSyntax,
+                                indexTemp,
+                                _factory.Binary(BinaryOperatorKind.Addition, indexTemp.Type, indexTemp, _factory.Literal(1)),
+                                isRef: false,
+                                indexTemp.Type));
```

Comment:
Before the first spread element, the compiler should emit constants indices rather than using an index variable.